esp-hal-common/src/i2c.rs

```diff
@@ -0,0 +1,968 @@
+        SCL: OutputPin<OutputSignal = OutputSignal> + InputPin<InputSignal = InputSignal>,
+    >(
+        i2c: T,
+        mut pins: Pins<SDA, SCL>,
+        frequency: u32,
+        system: &mut System,
+    ) -> Result<Self, SetupError> {
+        enable_peripheral(&i2c, system);
+
+        let mut i2c = I2C { peripheral: i2c };
+
+        pins.sda
+            .set_to_open_drain_output()
+            .enable_input(true)
+            .internal_pull_up(true)
+            .connect_peripheral_to_output(OutputSignal::I2CEXT0_SDA)
+            .connect_input_to_peripheral(InputSignal::I2CEXT0_SDA);
+
+        pins.scl
+            .set_to_open_drain_output()
+            .enable_input(true)
+            .internal_pull_up(true)
+            .connect_peripheral_to_output(OutputSignal::I2CEXT0_SCL)
+            .connect_input_to_peripheral(InputSignal::I2CEXT0_SCL);
+
+        i2c.peripheral.setup(frequency)?;
+
+        Ok(i2c)
+    }
+
+    /// Return the raw interface to the underlying peripheral
+    pub fn free(self) -> T {
+        self.peripheral
+    }
+}
+
+fn enable_peripheral<T: Instance>(i2c: &T, system: &mut System) {
+    // enable peripheral
+    #[cfg(feature = "esp32")]
+    match i2c.i2c_number() {
+        0 => {
+            system
+                .perip_clk_en
+                .modify(|_, w| w.i2c0_ext0_clk_en().set_bit());
+            system
+                .perip_rst_en
+                .modify(|_, w| w.i2c0_ext0_rst().clear_bit());
+        }
+        1 => {
+            system
+                .perip_clk_en
+                .modify(|_, w| w.i2c_ext1_clk_en().set_bit());
+            system
+                .perip_rst_en
+                .modify(|_, w| w.i2c_ext1_rst().clear_bit());
+        }
+        _ => panic!(), // will never happen
+    }
+    #[cfg(not(feature = "esp32"))]
+    match i2c.i2c_number() {
+        0 => {
+            system
+                .perip_clk_en0
+                .modify(|_, w| w.i2c_ext0_clk_en().set_bit());
+
+            // Take the I2C peripheral out of any pre-existing reset state
+            // (shouldn't be the case after a fresh startup, but better be safe)
+            system
+                .perip_rst_en0
+                .modify(|_, w| w.i2c_ext0_rst().clear_bit());
+        }
+        1 => {
+            cfg_if::cfg_if! {
+                if #[cfg(not(feature = "esp32c3"))] {
+                    system
+                    .perip_clk_en0
+                    .modify(|_, w| w.i2c_ext1_clk_en().set_bit());
+
+                    // Take the I2C peripheral out of any pre-existing reset state
+                    // (shouldn't be the case after a fresh startup, but better be safe)
+                    system
+                        .perip_rst_en0
+                        .modify(|_, w| w.i2c_ext1_rst().clear_bit());
+                } else {
+                    ()
+                }
+
+            }
+        }
+        _ => panic!(), // will never happen
```

Comment:
Nitpick, but this is probably "more correct":

```suggestion
        _ => unreachable!(), // will never happen
```